### ISSUE TITLE
poudriere fails to clean up

### ISSUE THREAD

**Issue by yurivict** [freebsd/poudriere#296]
I noticed this while troubleshooting issue #294

Here are the steps how I got to the problem:
- Begin from the empty poudriere/data directory
- Run some command like this 'poudriere bulk -j 10amd64 -p local -z workstation -t category/port'
- During the build of 'pkg' hit Ctrl-Z, and wait when pkg finishes. (because Ctrl-Z only stops poudriere, but not the current build itself)
- Change to /usr/local/poudriere/data/.m/10amd64-local-workstation/ and run command 'du -x -s *', enter directory '01' and runt the same command there
- Leave this directory
- Continue poudriere (fg) and then stop it (Ctrl-C) and wait
- Observe the messages, see below

I am not sure if this is intermittent, I only saw it once.

I can't imagine how can the file listing process with 'du' influence poudriere deleting things afterwards.

One might argue that this is not a real problem because it happened when someone did 'du' in the middle. But to me it means that there is some latent failure condition that doesn't normally get triggered, and might only come up during some unusual system load pattern or timing.

---- errors during clean up----

```
[00:09:00] ====>> Cleaning up
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/lib32/libc.so.7: Operation not permitted
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/lib32/libthr.so.3: Operation not permitted
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/lib32/librt.so.1: Operation not permitted
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/lib32/libcrypt.so.5: Operation not permitted
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/lib32: Directory not empty
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/bin/yppasswd: Operation not permitted
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/bin/opieinfo: Operation not permitted
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/bin/rlogin: Operation not permitted
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/bin/login: Operation not permitted
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/bin/opiepasswd: Operation not permitted
```

**Comment by yurivict** [comment]
I think that in the desktop environment there is this common problem that some agent may just open some directory when somebody happens to chdir into it, like this one /usr/local/libexec/gvfsd-trash from gvfs-1.20.3_2

So the problem may be that poudriere is missing -f in some 'umount' command? Usually umount fails when the target directory is opened and -f helps in such case. This is the common problem that many shell scripts suffer from (failure when some file or dir happens to be opened at the time).

**Comment by bdrewery** [comment]
Are you using NO_ZFS? That is the only way this would happen since in the ZFS and TMPFS cases it just unmounts the directory. I believe you actually are since that's the only plausible way #294 happens as well. It's possible your jail was created outside of ZFS which means it will not use ZFS going forward for that jail. If the file _/usr/local/etc/poudriere.d/jails/JAILNAME/fs_ does not have a ZFS dataset in it then it is not using ZFS.

In the NO_ZFS case it also runs `chflags -R noschg` to avoid the exact issue you reported. So I'm not sure how it happened. Did you modify something? Are you running Poudriere older than 3.0? That is when it was added.

Are you running Poudriere inside a jail? It needs `allow.chflags`.

**Comment by yurivict** [comment]
No, this option is commented out:

```
### NO ZFS
# To not use ZFS, define NO_ZFS=yes
#NO_ZFS=yes
```

**Comment by bdrewery** [comment]
And the fs file?

**Comment by yurivict** [comment]
Root filesystem is UFS.

**Comment by bdrewery** [comment]
So somehow the jail was created outside of ZFS which explains #294. Delete and recreate the jail to see if it fixes it.

**Comment by yurivict** [comment]
After I re-created the jail, now cleanup after an interrupted build fails even without chdir to the directories:
````
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/ref/../08/usr/ports/net-p2p/mldonkey/Makefile: Read-only file system
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/ref/../08/usr/ports/net-p2p/mldonkey/distinfo: Read-only file system
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/ref/../08/usr/ports/net-p2p/mldonkey/pkg-plist: Read-only file system
````

- It looks like poudriere was trying to cleanup on r/o mounts:
````
/usr/local/poudriere/jails/10amd64/usr/src on /usr/local/poudriere/data/.m/10amd64-local-workstation/ref/usr/src (nullfs, local, read-only)
/usr/local/poudriere/data/.m/10amd64-local-workstation/ref/usr/src on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/usr/src (nullfs, local, read-only)
devfs on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/dev (devfs, local, multilabel)
fdescfs on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/dev/fd (fdescfs)
procfs on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/proc (procfs, local)
linprocfs on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/compat/linux/proc (linprocfs, local)
tmpfs on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/wrkdirs (tmpfs, local)
/usr/local/poudriere/ports/local on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/usr/ports (nullfs, local, read-only)
/usr/local/poudriere/data/packages/10amd64-local-workstation/.building on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/packages (nullfs, local, read-only)
/usr/ports/distfiles on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/distfiles (nullfs, local)
/usr/local/poudriere/data/.m/10amd64-local-workstation/ref/var/db/ports on /usr/local/poudriere/data/.m/10amd64-local-workstation/08/var/db/ports (nullfs, local, read-only)
````

**Comment by yurivict** [comment]
It is maybe that poudriere can be caught with Ctrl-C signal at some point when the mounts are r/o and it doesn't make them writable.

**Comment by bdrewery** [comment]
Oh btw the "operation not permitted" issues earlier was due to a bug in 3.1.2 that was fixed in 3.1.3. https://github.com/freebsd/poudriere/wiki/release_notes

**Comment by bdrewery** [comment]
And probably your new finding. You should really upgrade, it's weeks old.

**Comment by yurivict** [comment]
Sorry I didn't realize poudriere was so badly outdated.
Updated to 3.1.4
All problems still there.

Please also note this additional problem while re-creating the jail (happened before too):
````
[root@yuri /usr/local/poudriere/yuri]# poudriere jails -d -j 10amd64
[00:00:00] ====>> Removing 10amd64 jail... done
[root@yuri /usr/local/poudriere/yuri]# poudriere jail -c -j 10amd64 -v 10.1-RELEASE -a amd64
[00:00:00] ====>> Creating 10amd64 fs... done
mkdir: /usr/local/poudriere/jails/10amd64: No such file or directory
[00:00:00] ====>> FREEBSD_HOST from config invalid; defaulting to http://ftp.freebsd.org
fetch: /usr/local/poudriere/jails/10amd64/fromftp/MANIFEST: open(): No such file or directory
fetch: /usr/local/poudriere/jails/10amd64/fromftp/MANIFEST: open(): No such file or directory
[00:00:01] ====>> Error: Failed to fetch from http://ftp.freebsd.org/pub/FreeBSD/releases/amd64/amd64/10.1-RELEASE/MANIFEST
[00:00:01] ====>> Error while creating jail, cleaning up.
[00:00:01] ====>> Removing 10amd64 jail... done
````
This command fixes it:
````
mkdir /usr/local/poudriere/jails/10amd64
````

Ctrl-C cleanup is still broken:
````
[00:00:14] ====>> Starting/Cloning builders
^C[00:00:35] ====>> Error: Signal SIGINT caught, cleaning up and exiting
[00:00:35] ====>> Cleaning up
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/share: Directory not empty
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr: Directory not empty
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/01: Directory not empty
/usr/local/poudriere/data/.m/10amd64-local-workstation/02/usr/sbin/chown hardlink: unable to link to /usr/local/poudriere/data/.m/10amd64-local-workstation/02/usr/bin/chgrp: No such file or directory
/usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/sbin/chown hardlink: unable to link to /usr/local/poudriere/data/.m/10amd64-local-workstation/01/usr/bin/chgrp: No such file or directory
/usr/local/poudriere/data/.m/10amd64-local-workstation/01/root/.profile hardlink: unable to link to /usr/local/poudriere/data/.m/10amd64-local-workstation/01/.profile: No such file or directory
/usr/local/poudriere/data/.m/10amd64-local-workstation/01/root/.cshrc hardlink: unable to link to /usr/local/poudriere/data/.m/10amd64-local-workstation/01/.cshrc: No such file or directory
/usr/local/poudriere/data/.m/10amd64-local-workstation/02/root/.profile hardlink: unable to link to /usr/local/poudriere/data/.m/10amd64-local-workstation/02/.profile: No such file or directory
/usr/local/poudriere/data/.m/10amd64-local-workstation/02/root/.cshrc hardlink: unable to link to /usr/local/poudriere/data/.m/10amd64-local-workstation/02/.cshrc: No such file or directory
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/02/usr: Directory not empty
rm: /usr/local/poudriere/data/.m/10amd64-local-workstation/02: Directory not empty
[00:01:49] ====>> Umounting file systems
````

And 3.1.4 still eats up the root disk space (from another issue)

**Comment by khera** [comment]
I am observing cleanup error with 3.1.15 now as well. I had one interrupted build yesterday, and since then every time I run `bulk` build, it gives cleanup errors. I rebooted, and I manually unmounted all these file systems, and I deleted any spurious ZFS clones. This failure persists.

```
[00:00:24] ====>> Cleaning up
[00:00:24] ====>> Unmounting file systems
mdconfig: ioctl(/dev/mdctl): Device busy
umount: unmount of /usr/local/poudriere/data/.m/11amd64-default/ref/usr/src failed: Device busy
umount: unmount of /usr/local/poudriere/data/.m/11amd64-default/ref/rescue failed: Device busy
umount: unmount of /usr/local/poudriere/data/.m/11amd64-default/ref/usr/src failed: Device busy
umount: unmount of /usr/local/poudriere/data/.m/11amd64-default/ref/rescue failed: Device busy
cannot unmount '/usr/local/poudriere/data/.m/11amd64-default/ref': Device busy
rmdir: /usr/local/poudriere/data/.m/11amd64-default/ref: Device busy
rm: /usr/local/poudriere/data/.m/11amd64-default/ref/../ref: Device busy
rm: /usr/local/poudriere/data/.m/11amd64-default/ref/../: Invalid argument
```

It also leaves a bunch of md devices hanging around, which I also manually remove between runs.
```
# mdconfig -l -v
md0     swap       64M
md1     swap     4096M
md2     swap     4096M
md3     swap     4096M
md4     swap     4096M
```

Device md0 is a part of my base system. 1-4 are made by poudriere and not cleaned up.

Basically I have to manually run:

```
df | grep /usr/local/poudriere/data/.m/ | awk '{print $6}' | xargs umount # repeat until all are unmounted
zfs destroy zfspool/poudriere/jails/11amd64-default-ref
mdconfig -d -u 1 # repeat for each leftover device
```

after every run.

When I do a `bulk -n` dry run, the cleanup is different:

```
[00:00:19] [Dry Run] ====>> Cleaning up
[00:00:19] [Dry Run] ====>> Unmounting file systems
mdconfig: ioctl(/dev/mdctl): Device busy
```

And I only need to remove the leftover md device.

This was not happening until yesterday when two things happened: 1) I had an interrupted poudriere bulk run with lots of ports and 2) poudriere itself was upgraded from 3.1.14_3 to 3.1.15. I do not know which of these is the cause.

**Comment by khera** [comment]
The above failure to clean up is still happening to me for 3.1.16. I reverted to 3.1.14 from the quarterly package collection, and the cleanup happens properly now. My build host is running FreeBSD 11.0.

**Comment by bdrewery** [comment]
I'll look into it.  However I've been tempted to remove the MDMFS support entirely as it is bitrotting.  The `USE_TMPFS` feature is very supported and also avoids extra overhead of UFS in memory.  Do you have a reason to use MDMFS over USE_TMPFS or was it just not knowing about the tmpfs option?

When poudriere started tmpfs was very unstable and immature in FreeBSD, but it is quite mature and stable now.

**Comment by bdrewery** [comment]
I do see the bug with MDMFS support in 3.1.15/3.1.16. I will get it fixed for 3.1.17 for sure.

**Comment by bdrewery** [comment]
@khera bc98ecb7450bd1b7f786d31ef8ee4c34d6f0f340 fixes it and 2f3c947288a2a107bf1334c1c900e00caa8193e6 makes `jail -k` work better in this case as well. It will be in 3.1.17.

**Comment by khera** [comment]
I have no real reason to use MDMFS over TMPFS and I do not know why I set it up that way. Perhaps I read some recommendation on some random web page to do it that way.

I'll reconfigure. Thanks for the update.

**Comment by bdrewery** [comment]
Pretty sure this is fixed by #654 and other updates.
